### ISSUE TITLE
CONSISTENCY: Lack of symmetry in warnings produced by sum2  

### ISSUE THREAD

**Issue by PeteHaitch** [HenrikBengtsson/matrixStats#98]
Getting the `mode` "wrong" in `matrixStats::sum2()` returns a warning in some cases but not others. I'm not certain this is really the responsibility of **matrixStats**, but I thought I'd raise it. Arguably the developer/user should check that the output is as expected, but since `matrixStats::sum2()` tries in some cases but not others to warn the user of unexpected output I thought I'd raise it as a potential issue with consistency.

``` r
# Integer vector that produces overflow. matrixStats::sum2() warns if mode is 
# 'incorrect', so could wrap in try() to handle warning case
x_overflow <- c(1L, .Machine$integer.max)
# "Wrong" mode -> warning
str(matrixStats::sum2(x_overflow))
#> Warning in matrixStats::sum2(x_overflow): Integer overflow. Use sum2(...,
#> mode="numeric") to avoid this.
#>  int NA
# "Wrong" mode -> warning
str(matrixStats::sum2(x_overflow, mode = "integer"))
#> Warning in matrixStats::sum2(x_overflow, mode = "integer"): Integer
#> overflow. Use sum2(..., mode="numeric") to avoid this.
#>  int NA
# "Correct" mode -> correct output
str(matrixStats::sum2(x_overflow, mode = "double"))
#>  num 2.15e+09

# A vector that was expected to be integer but isn't (e.g. returned by another 
# routine). matrixStats::sum2() doesn't warn if mode is 'incorrect', so can't 
# wrap in try() to handle warning case
x_thought_was_int <- c(1, 1.7)
# "Correct" mode -> correct output
str(matrixStats::sum2(x_thought_was_int))
#>  num 2.7
# "Wrong" mode -> **no warning** and likely unexpected output
str(matrixStats::sum2(x_thought_was_int, mode = "integer"))
#>  int 2
# "Correct" mode -> correct output
str(matrixStats::sum2(x_thought_was_int, mode = "double"))
#>  num 2.7
```

A user who *truly* wants an integer return value from summing a numeric/double vector could still use `suppressWarnings(matrixStats::sum2(x_thought_was_int, mode = "integer"))`.

**Comment by HenrikBengtsson** [comment]
Is the idea that `sum2(<double>, mode = "integer")` should always give a warning (or possibly even be an error), because it's very likely a user error?  If really intended, one should probably use `as.integer(sum2(<double>))`.

**Comment by PeteHaitch** [comment]
I think it should at least trigger a warning. Your argument that one should probably use the explicit `as.integer(sum2(<double>))` rather than relying on (possibly silent) coercion does sway me that signalling an error may be appropriate in this case

**Comment by HenrikBengtsson** [comment]
Implemented in commit eb2ff36:
```r
> matrixStats::sum2(c(1,2), mode = "integer")
[1] 3
Warning message:
In matrixStats::sum2(c(1, 2), mode = "integer") :
  sum2(x, mode = "integer") called with typeof(x) == "double"; did you mean to use as.integer(sum2(x))?
```

Leaving it as a warning for now though one could argue for an error as well.
